scalding-core/src/main/scala/com/twitter/scalding/graph/Graph.scala

```diff
@@ -0,0 +1,231 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package com.twitter.scalding.graph
+
+import com.twitter.scalding.TypedPipe
+
+import scala.reflect.ClassTag
+
+class Graph[T: ClassTag, S, Q](inputEdges: TypedPipe[Edge[T, S]], inputVertices: TypedPipe[Vertex[T, Q]])(implicit ord: Ordering[T]) {
+  def edges: TypedPipe[Edge[T, S]] = inputEdges
+  def vertices: TypedPipe[Vertex[T, Q]] = inputVertices
+
+  /**
+   * Returns a TypedPipe of edges with joined Vertex attributes.
+   */
+  def triplets: TypedPipe[EdgeTriplet[T, S, Q]] =
+    edges
+      .groupBy(_.source)
+      .join(vertices.groupBy(_.id))
+      .toTypedPipe
+      .map{ case (source, (edge, packet)) => (edge.dest, (edge, packet)) }
+      .join(vertices.groupBy(_.id))
+      .values
+      .map{ case ((edge, sourceVertex), destVertex) => EdgeTriplet(sourceVertex, destVertex, edge) }
+
+  /**
+   * Left join vertices with the graph vertices and generate a new Graph
+   */
+  def leftJoinVertices[U, VD2](other: TypedPipe[Vertex[T, U]])(mapFunc: (T, Q, Option[U]) => VD2): Graph[T, S, VD2] = {
+    val newVertices = vertices
+      .groupBy(_.id)
+      .leftJoin(other.groupBy(_.id))
+      .toTypedPipe
+      .map{ case (id, (vertex, data)) => Vertex(id, mapFunc(id, vertex.attr, data.map(_.attr))) }
+
+    new GraphUnfilteredEdges(edges, newVertices)
+  }
+
+  /**
+   * Inner join vertices with the graph vertices and generate a new Graph
+   */
+  def joinVertices[U, VD2](other: TypedPipe[Vertex[T, U]])(mapFunc: (T, Q, U) => VD2): Graph[T, S, VD2] = {
+    val newVertices = vertices
+      .groupBy(_.id)
+      .join(other.groupBy(_.id))
+      .toTypedPipe
+      .map{ case (id, (vertex, data)) => Vertex(id, mapFunc(id, vertex.attr, data.attr)) }
+
+    new Graph[T, S, VD2](edges, newVertices)
+  }
+
+  def mapVertices[A](map: Vertex[T, Q] => A): Graph[T, S, A] =
+    new Graph[T, S, A](edges, vertices.map{ vertex => Vertex(vertex.id, map(vertex)) })
+
+  def mapEdges[A](map: Edge[T, S] => A): Graph[T, A, Q] =
+    new Graph[T, A, Q](edges.map{ edge => edge.copy(attr = map(edge)) }, vertices)
+
+  def mapTriplets[A, B](map: EdgeTriplet[T, S, Q] => EdgeTriplet[T, A, B]): Graph[T, A, B] = {
+    val newTriplets = triplets.map(map)
+
+    new Graph[T, A, B](
+      newTriplets.map(_.edge),
+      newTriplets.flatMap(trip => List(trip.source, trip.dest)).distinct(Ordering.by(_.id)))
+  }
+
+  /**
+   * Collect only neighbor ids.
+   * Optionally sort by Vertex
+   */
+  def collectNeighborIds(sortNeighbors: Boolean = true): Graph[T, S, Neighbors[T]] = {
```

Comment:
Thats a good point.  I put sort there because Neighbors has two subclasses, one called Sorted and Unsorted.  This code is coming from a branch where I was doing lots of work with Neighbor intersection where being sorted has a huge advantage and have a type to confirm sorted was nice check on that.

I can also create a sort function on UnsortedNeighbors which may be more explicit.